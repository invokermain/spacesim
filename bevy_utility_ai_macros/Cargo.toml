[package]
name = "bevy_utility_ai_macros"
version = "0.0.1"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
bevy = "0.10"
syn = { version = "1", features = ["extra-traits", "full", "fold"] }
quote = "1"
proc-macro2 = "1"
proc-macro-error = "1"
