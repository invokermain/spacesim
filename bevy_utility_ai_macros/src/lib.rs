#[cfg(not(target_family = "wasm"))]
#[proc_macro_error::proc_macro_error]
#[proc_macro_attribute]
pub fn input_system(
    args: proc_macro::TokenStream,
    input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    inner::input_system_core(args.into(), input.into()).into()
}

#[cfg(not(target_family = "wasm"))]
mod inner {
    use proc_macro2::TokenStream;
    use proc_macro_error::abort;
    use quote::quote;
    use syn::{parse2, FnArg, ItemFn};

    pub(crate) fn input_system_core(_: TokenStream, input: TokenStream) -> TokenStream {
        let item_fn = match parse2::<ItemFn>(input) {
            Ok(syntax_tree) => syntax_tree,
            Err(error) => return error.to_compile_error(),
        };

        let name = item_fn.sig.ident;

        let inputs: Vec<_> = item_fn
            .sig
            .inputs
            .iter()
            .map(|input| match input {
                FnArg::Receiver(_) => abort!(input, "Input function cannot have self"),
                FnArg::Typed(arg) => {
                    let arg_name = match arg.pat.as_ref() {
                        syn::Pat::Ident(ident) => &ident.ident,
                        _ => abort!(arg, "Expected Ident?"),
                    };
                    let arg_type = match arg.ty.as_ref() {
                        syn::Type::Reference(reference) => match &reference.elem.as_ref() {
                            syn::Type::Path(path) => path
                                .path
                                .segments
                                .last()
                                .unwrap_or_else(|| abort!(path.path, "What?")),
                            _ => abort!(reference.elem, "Expected a Component"),
                        },
                        _ => abort!(arg.ty, "Expected the parameter type to be a reference"),
                    };
                    (arg_name, arg_type)
                }
            })
            .collect();

        let types = inputs.iter().map(|(_, t)| t);
        let names = inputs.iter().map(|(n, _)| n);

        let body = item_fn.block;

        let output = quote! {

            fn #name(mut query: bevy::prelude::Query<(&mut bevy_utility_ai::AIMeta #(, &#types)*)>) {
                let key = #name as usize;
                for (mut ai_meta #(, #names)*) in query.iter_mut() {
                    let score = #body;
                    let mut entry = ai_meta.input_scores.entry(key).or_insert(f32::NEG_INFINITY);
                    *entry = score;
                }
            }
        };

        output.into()
    }
}

// --------------------------------------------------------------------------------------
// The WASM implementation of the above that cannot use proc_macro2 and related features,
// the only difference should be in error handling.
// --------------------------------------------------------------------------------------
#[cfg(target_family = "wasm")]
#[proc_macro_attribute]
pub fn input_system(
    args: proc_macro::TokenStream,
    input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let item_fn = match parse::<ItemFn>(input) {
        Ok(syntax_tree) => syntax_tree,
        Err(error) => panic!(),
    };

    let name = item_fn.sig.ident;

    let inputs: Vec<_> = item_fn
        .sig
        .inputs
        .iter()
        .map(|input| match input {
            FnArg::Receiver(_) => panic!(),
            FnArg::Typed(arg) => {
                let arg_name = match arg.pat.as_ref() {
                    syn::Pat::Ident(ident) => &ident.ident,
                    _ => panic!(),
                };
                let arg_type = match arg.ty.as_ref() {
                    syn::Type::Reference(reference) => match &reference.elem.as_ref() {
                        syn::Type::Path(path) => {
                            path.path.segments.last().unwrap_or_else(|| panic!())
                        }
                        _ => panic!(),
                    },
                    _ => panic!(),
                };
                (arg_name, arg_type)
            }
        })
        .collect();

    let types = inputs.iter().map(|(_, t)| t);
    let names = inputs.iter().map(|(n, _)| n);

    let body = item_fn.block;

    let output = quote! {

        fn #name(mut query: bevy::prelude::Query<(&mut bevy_utility_ai::AIMeta #(, &#types)*)>) {
            let key = #name as usize;
            for (mut ai_meta #(, #names)*) in query.iter_mut() {
                let score = #body;
                let mut entry = ai_meta.input_scores.entry(key).or_insert(f32::NEG_INFINITY);
                *entry = score;
            }
        }
    };

    output.into()
}

#[cfg(test)]
mod tests {
    use crate::inner::input_system_core;
    use quote::quote;

    #[test]
    fn first() {
        let before = quote! {
            fn utility_input_low(some_data: &SomeData) -> f32 {
                some_data.val
            }
        };

        input_system_core(quote!(), before.into());
    }
}
