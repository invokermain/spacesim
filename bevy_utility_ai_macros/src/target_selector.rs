use proc_macro::TokenStream;
use quote::{quote, ToTokens};
use syn::{Error, FnArg, ItemFn, Pat, Type};

pub(crate) fn target_selector(
    _: TokenStream,
    input: TokenStream,
) -> Result<TokenStream, Error> {
    let item_fn = match syn::parse::<ItemFn>(input) {
        Ok(ast) => ast,
        Err(err) => return Err(err),
    };

    let name = item_fn.sig.ident;
    let body = item_fn.block;

    if item_fn.sig.inputs.len() != 1 {
        return Err(Error::new_spanned(
            item_fn.sig.inputs.into_token_stream(),
            "Expected only one argument".to_string(),
        ));
    }

    let input = match &item_fn.sig.inputs[0] {
        FnArg::Receiver(_) => panic!(),
        FnArg::Typed(typed_input) => typed_input,
    };

    let input_ident = match input.pat.as_ref() {
        Pat::Ident(ident) => ident,
        _ => {
            return Err(Error::new_spanned(
                input,
                "expected an identity here!".to_string(),
            ))
        }
    };

    let input_ty = match input.ty.as_ref() {
        Type::Path(path) => path.to_token_stream(),
        _ => {
            return Err(Error::new_spanned(
                input,
                "expected an type path here!".to_string(),
            ));
        }
    };

    let output = quote! {
        fn #name(
            mut q_subject: bevy::prelude::Query<(bevy::prelude::Entity, &mut bevy_utility_ai::AIMeta)>,
            #input_ident: #input_ty,
            ai_definitions: bevy::prelude::Res<bevy_utility_ai::AIDefinitions>,
        ) {
            let target_selector_key = #name as usize;
            for (entity, mut ai_meta) in q_subject.iter_mut() {
                let targets = #body;


                if let Some(val) = ai_meta.targeted_input_targets.get_mut(&key) {
                    *val = targets;
                } else {
                    ai_meta.targeted_input_targets.insert(key, targets);
                }
            }
        }
    };

    Ok(output.into())
}
