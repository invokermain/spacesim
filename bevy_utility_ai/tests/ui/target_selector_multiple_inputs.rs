use bevy_utility_ai_macros::target_selector;

#[target_selector]
fn my_target_selector(q_a: Query<A>, q_a: Query<B>) -> Vec<Entity> {
    q_targets.iter().map(|x| x.0).collect();
}

fn main() {}
