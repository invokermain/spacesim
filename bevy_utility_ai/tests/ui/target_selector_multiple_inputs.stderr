error: Expected only one argument
 --> tests/ui/target_selector_multiple_inputs.rs:4:23
  |
4 | fn my_target_selector(q_a: Query<A>, q_a: Query<B>) -> Vec<Entity> {
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
